langchain/agents/openai_functions_multi_agent/base.py

```diff
@@ -0,0 +1,348 @@
+            else:
+                tool_input = _tool_input
+
+            content_msg = "responded: {content}\n" if message.content else "\n"
+            log = f"\nInvoking: `{function_name}` with `{tool_input}`\n{content_msg}\n"
+            _tool = _FunctionsAgentAction(
+                tool=function_name,
+                tool_input=tool_input,
+                log=log,
+                message_log=[message],
+            )
+            final_tools.append(_tool)
+        return final_tools
+
+    return AgentFinish(return_values={"output": message.content}, log=message.content)
+
+
+class OpenAIMultiFunctionsAgent(BaseMultiActionAgent):
+    """An Agent driven by OpenAIs function powered API.
+
+    Args:
+        llm: This should be an instance of ChatOpenAI, specifically a model
+            that supports using `functions`.
+        tools: The tools this agent has access to.
+        prompt: The prompt for this agent, should support agent_scratchpad as one
+            of the variables. For an easy way to construct this prompt, use
+            `OpenAIFunctionsAgent.create_prompt(...)`
+    """
+
+    llm: BaseLanguageModel
+    tools: Sequence[BaseTool]
+    prompt: BasePromptTemplate
+
+    def get_allowed_tools(self) -> List[str]:
+        """Get allowed tools."""
+        return list([t.name for t in self.tools])
+
+    @root_validator
+    def validate_llm(cls, values: dict) -> dict:
+        if not isinstance(values["llm"], ChatOpenAI):
+            raise ValueError("Only supported with ChatOpenAI models.")
+        return values
+
+    @root_validator
+    def validate_prompt(cls, values: dict) -> dict:
+        prompt: BasePromptTemplate = values["prompt"]
+        if "agent_scratchpad" not in prompt.input_variables:
+            raise ValueError(
+                "`agent_scratchpad` should be one of the variables in the prompt, "
+                f"got {prompt.input_variables}"
+            )
+        return values
+
+    @property
+    def input_keys(self) -> List[str]:
+        """Get input keys. Input refers to user input here."""
+        return ["input"]
+
+    @property
+    def functions(self) -> List[dict]:
+        enum_vals = [t.name for t in self.tools]
+        tool_selection = {
+            "name": "ToolSelection",
+            "description": "A list of actions to take.",
+            "parameters": {
+                "title": "ToolSelection",
+                "description": "A list of actions to take.",
+                "type": "object",
+                "properties": {
+                    "actions": {
+                        "title": "Actions",
+                        "type": "array",
+                        "items": {
+                            "title": "ToolCall",
+                            "type": "object",
+                            "properties": {
+                                "tool_name": {
+                                    "title": "Tool Name",
```

Comment:
not sure if relevant but i think having descriptions can help with correct usage. but i noticed that its based on enum of tool name rather than the tools properties.